### ISSUE TITLE
[tox4] Differing discovered environments for factors

### ISSUE THREAD

**Issue by jugmac00** [tox-dev/tox#2039]
caused by these lines

```
commands_pre =
    py27,py35: {envbindir}/buildout -nc {toxinidir}/buildout4.cfg buildout:directory={envdir} buildout:develop={toxinidir} install test
    !py27-!py35: {envbindir}/buildout -nc {toxinidir}/buildout.cfg buildout:directory={envdir} buildout:develop={toxinidir} install test
```

in repo 
```
git clone git@github.com:zopefoundation/Products.PythonScripts.git
```

```
tox4dev config
...
[testenv:py27-py35]
type = VirtualEnvRunner
set_env = PIP_DISABLE_PIP_VERSION_CHECK=1
base = testenv
runner = virtualenv
description = 
depends = 
env_name = py27-py35
env_dir = /home/jugmac00/All/output_zope/zopefoundation/Products.PythonScripts/.tox/4/py27-py35
env_tmp_dir = /home/jugmac00/All/output_zope/zopefoundation/Products.PythonScripts/.tox/4/py27-py35/tmp
platform = 
pass_env =
  CURL_CA_BUNDLE
  LANG
  LANGUAGE
  LD_LIBRARY_PATH
  PIP_*
  REQUESTS_CA_BUNDLE
  SSL_CERT_FILE
  TERM
  TMPDIR
  VIRTUALENV_*
  http_proxy
  https_proxy
  no_proxy
parallel_show_output = False
recreate = False
commands_pre = # Exception: HandledError("replace failed in py27-py35.commands_pre with ValueError('conflicting factors py27, py35 in py27-py35')")
commands = # Exception: HandledError("replace failed in py27-py35.commands with ValueError('conflicting factors py27, py35 in py27-py35')")
commands_post = 
change_dir = /home/jugmac00/All/output_zope/zopefoundation/Products.PythonScripts
ignore_errors = False
ignore_outcome = False
skip_install = True
package = skip
base_python = # Exception: ValueError('conflicting factors py27, py35 in py27-py35')
env_site_packages_dir = # Exception: ValueError('conflicting factors py27, py35 in py27-py35')
env_bin_dir = # Exception: ValueError('conflicting factors py27, py35 in py27-py35')
env_python = # Exception: ValueError('conflicting factors py27, py35 in py27-py35')
deps =
  setuptools < 52
  zc.buildout
system_site_packages = False
always_copy = False
download = False
...
```

![Screenshot from 2021-04-19 18-02-01](https://user-images.githubusercontent.com/9895620/115267359-6256de00-a139-11eb-9acf-d9d07984f741.png)

**Comment by gaborbernat** [comment]
This is on purpose, per https://github.com/tox-dev/tox/blob/8d6a32436dc4f79ff2f9241db0603bdb47f1744c/src/tox/config/__init__.py#L666-L667. We were warning in tox 3 that this will become an error in tox 4. 🤔 Maybe we still want to support though, ignore_basepython_conflict flag as part of this ticket to allow users to bail out.

**Comment by jugmac00** [comment]
I spent some time to understand this issue or where there is a problem at all.

Is it just that we do not want to support...
```
py27,py35: {envbindir}/buildout -nc {toxinidir}/buildout4.cfg buildout:directory={envdir} buildout:develop={toxinidir} install test
```
but users should write each factor on a single line like...
```
py27: {envbindir}/buildout -nc {toxinidir}/buildout4.cfg buildout:directory={envdir} buildout:develop={toxinidir} install test
py35: {envbindir}/buildout -nc {toxinidir}/buildout4.cfg buildout:directory={envdir} buildout:develop={toxinidir} install test
```
?

When I run tox3 for the mentioned package, I do not get any warning - not for `--showconfig` and not for running the testenvs.

I even set a breakpoint above the line you showed me, but it was not hit.

For tox4 on the other hand, I **only** see a problem when running `tox4 config` - but I see 0 errors when running all envs...

```
  lint: OK (9.89=setup[7.42]+cmd[0.01,0.37,0.56,0.52,0.86,0.15] seconds)
  py27: OK (81.73=setup[1.42]+cmd[77.86,2.44] seconds)
  py35: OK (74.37=setup[1.97]+cmd[69.49,2.91] seconds)
  py36: OK (69.83=setup[3.76]+cmd[63.66,2.41] seconds)
  py37: OK (73.46=setup[3.65]+cmd[67.06,2.75] seconds)
  py38: OK (79.32=setup[3.57]+cmd[73.00,2.74] seconds)
  py39: OK (132.85=setup[4.22]+cmd[126.04,2.59] seconds)
  coverage: OK (77.35=setup[5.04]+cmd[67.90,0.00,3.70,0.41,0.30] seconds)
  congratulations :) (598.86 seconds)
```

**Comment by gaborbernat** [comment]
The problem here is the environment ``py27-py35``. Here you have two python specifications in the name, so which one should we pick?

**Comment by gaborbernat** [comment]
Comparing the output of `tox -av` and `tox4 -av` I see that tox 4 actually also discovers `py27-py35` as a target environment. This should not happen. We need to understand why tox 4 does discover this env, and why tox 3 does not. It must do with the conditional line ` !py27-!py35:` though not sure why this does not cause a problem for tox 3.

**Comment by jugmac00** [comment]
I spent some more time debugging this...

And I pinned the problem down to the first regex in the following function...

https://github.com/tox-dev/tox/blob/072ca8d9a66ccaa623a4b20ad6fdc2498d006e8c/src/tox/config/loader/ini/factor.py#L67-L76

There a value like `'!py36-!py37'`  (or `'py36,py37'` or `' py37-django' `) only gets split at commas, but not on dashes.

When I update the regex to also split on dashes, this one problem is fixed, but certainly all examples like `'py37-django'` get out of hand.

I have a hard time to come up with a proper solution. I think at one point we would need to check whether in a combined factor (combined with a dash) all parts are Python versions or only one of them.

Or have a look how this is solved in `tox3` as you said ;-)

**Comment by gaborbernat** [comment]
Definitely, we first need to understand how this is solved in tox 3 and then we can go from there 🤔 We might solve it differently in tox 4, but it's important we understand first why this is not an issue in the past.

**Comment by jugmac00** [comment]
Oh my... I so was on the wrong track...

If my assumptions are right that envs can only be defined in two ways:
- envlist in the tox section (eg `envlist = py37, 38`)
- section header of testenvs (eg `[testenv:py39`])
... then in 

https://github.com/tox-dev/tox/blob/072ca8d9a66ccaa623a4b20ad6fdc2498d006e8c/src/tox/config/source/ini.py#L101-L113

the last `yield from` is the problem. It scans the contents of the key value pairs within sections, which at least for `tox.ini` is not what we want - no clue about `setup.cfg` or others - need to lookup the syntax.

Ah, interesting...

When I remove the last three lines, the following test fails...

https://github.com/tox-dev/tox/blob/072ca8d9a66ccaa623a4b20ad6fdc2498d006e8c/tests/config/test_main.py#L26-L42

With the 3 lines from above removed, `other` does not get found. But... `other` does not get found in `tox3` anyway!!

```
[tox]
envlist = py38, py37
[testenv]
deps = 1
    other: 2
[testenv:magic]
```

```
❯ tox -av -c tox.xxx.ini 
using tox.ini: /home/jugmac00/Projects/tox/tox.xxx.ini (pid 23096)
using tox-3.23.0 from /home/jugmac00/opt/tox/lib/python3.8/site-packages/tox/__init__.py (pid 23096)
default environments:
py38  -> [no description]
py37  -> [no description]

additional environments:
magic -> [no description]
```

```
❯ tox4dev -av -c tox.xxx.ini  # current head
default environments:
py38  -> [no description]
py37  -> [no description]

additional environments:
other -> [no description]
magic -> [no description]
```

```
❯ tox4dev -av -c tox.xxx.ini  # the last three lines removed
default environments:
py38  -> [no description]
py37  -> [no description]

additional environments:
magic -> [no description]
```

So, looks like you changed the env discovery logic from tox3 to tox4 in a sense that in tox4 within a testenv section, ie in a value for a key, it should be scanned for envs!! Was this intentional?? :crossed_fingers: for no, then the fix would be easy :-)

The second failing test is `test_factor_config_no_env_list_creates_env` and here it is basically the same problem:

```
[tox]
envlist = py37-{django15,django16}
[testenv]
deps =
    pytest
    django15: Django>=1.5,<1.6
    django16: Django>=1.6,<1.7
    py36: unittest2
```

Both tox4 with my suggested removal and tox3 show only the specified envs, current tox4 head shows additionally py36.


I am big fan of "explicit is better than implicit" => no implicit defining a testenv by only adding a factor somewhere; tox config syntax is already pretty elaborate - both for the users and for us to maintain. Let's not make it even more complicated.

**Comment by gaborbernat** [comment]
> If my assumptions are right that envs can only be defined in two ways:
> 
> * envlist in the tox section (eg `envlist = py37, 38`)
> * section header of testenvs (eg `[testenv:py39`])

It also can be defined as a factor literal for any of the configuration values. This is intentional. This not showing up under tox 3 was a limitation of that implementation. In the above example, tox py36 is explicitly defined under deps, so it's a non-default explicit environment. I feel we should only ignore negative expressions when doing such extension.

**Comment by gaborbernat** [comment]
You've mentioned you've been looking into this, so to mark this I'll assign this ticket to you @jugmac00

**Comment by gaborbernat** [comment]
So on a deeper look, the issue is this factor negation line.

```
!py27-!py35: {envbindir}/buildout -nc {toxinidir}/buildout.cfg buildout:directory={envdir} buildout:develop={toxinidir} install test
```

This line is bugged in tox 3. It really should be `!py27,!py35` to cover the intent, though `-` happens to work (by chance and is fixed in tox 4). Note the line above it uses the correct syntax for the affirmative selection. 

The `,` character in factor expression is loosely translating to `or` operator: 

- `py27,py35:` means use this configuration if the env contains factors py27 or py35. 
- `!py27,!py35:` means use this configuration if the env contains factors not of py27 or not of py35.
- `!py27-!py35:` means use this configuration if the environment name is a factor expression where the first element is not py27 and the second element is not py35. E.g. such env would be `py39-magic`. 

Let's unroll this final form. You specify a factor of the form `!py27-!py35`. This allows us to conclude that it's possible to have environments in the form of `a-b`. Furthermore, the first position can be `py27`, and in position two can be `py35` (otherwise, the negative filtering would not do anything and could be omitted). Therefore, we can conclude that it's possible to have the tox environment `py27-py35`, which we generate in the possible environments list. This is what you see in `tox4 l` or `tox4 c` (when you enlist non-default target environments).

In conclusion, tox 4 does here the right thing. tox 3 is buggy though and wrongfully accepts `!py27-!py35:` instead of `!py27,!py35:`.
